docs/proposals/11-cluster-health.md

```diff
@@ -0,0 +1,298 @@
+
+##### Possible Actions
+
+- When this condition transitions from status `false` to `true` (i.e. `etcd` transitions from being unreachable to reachable), the dependent `pods` could be deleted but only if they are in `CrashLoopBackoff`. This is to make sure the dependent `pods` recover as fast as possible from an `etcd` down time. This could be done by the restarter part of the current `root` [command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/root.go) of the [dependency-watchdog](https://github.com/gardener/dependency-watchdog) but re-factored as a separate restarter component which restarts (deletes) some targets `pods` based on some configured (via command-line argument, environment variables, config files etc.) conditions in the `Cluster` resource but only if the `pods` are in `CrashLoopBackoff`. 
+
+#### KubeAPIServerReachableInternally
+
+The `KubeAPIServerReachableInternally` condition type can be implemented (both the monitoring and the remedial action parts) similar to the [`ETCDReachable`](#etcdreachable) condition type.
+
+This condition may not be directly usable to take action.
+However, it might be useful in avoiding taking some pointless (and potentially harmful) actions.
+For example, if `KubeAPIServerReachableInternally` is `false` then we can avoid [scaling down the `kube-controller-manager`](#possible-actions) even if `KubeAPIServerReachableExternally` is `false`.
+
+#### UnhealthyNodesThresholdReached
+
+A component can monitor the fraction of the `NotReady` nodes in a zone and update the conditions with the `UnhealthyNodesThresholdReached` condition type.
+The list of affected zones could be maintained in the [`Message`](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L168).
+
+##### Possible Actions
+
+- If necessary, `kube-controller-manager` could be scaled down as described [above](#possible-actions).
+- Some resource creations (say, `pods`) could be throttled/rate-limited.
+- The `machine-controller-manager` could be either scaled down as described [above](#possible-actions) or be enhanced to avoid (or ramp down) deletion of machines from the affected zones (or `machinedeployments`).
+- Gardener can defer any disruptive updates until later.
+
+### Condition Types of future relevance
+
+This is just a sample set of condition types that might be of future relevance. It is not an exhaustive list.
+The suggested possible actions for these condition types are also just samples and not an exhaustive list.
+
+#### InternalNetworkHealthy
+
+If there is a way to detect problems in the cluster network that is severe enough to affect any significant portion of the cluster then this could be encoded in a component that updates the conditions with the `InternalNetworkHealthy` condition type.
```

Comment:
You file a proposal and some of your sentences are in form of:
> If there is a way to ..., then we could ...

I guess we would need to be more concrete and precise.

Reply:
> You file a proposal and some of your sentences are in form of:
> 
> > If there is a way to ..., then we could ...
> 
> I guess we would need to be more concrete and precise.

When a proposal is to change the way the cluster health conditions are published (the current way is  only suitable for external consumption) to make it more suitable for internal consumption, I can only make arguments about how making it possible for to consume internally would be useful in a host of real and hypothetical cases. The specific [current](https://github.com/gardener/gardener/pull/1797/files/0066744bd405637094f0df87e8671dce31b08002#diff-dc2b4cab766fb3ad06a4e681a236b570R177) and [hypothetical](https://github.com/gardener/gardener/pull/1797/files/0066744bd405637094f0df87e8671dce31b08002#diff-dc2b4cab766fb3ad06a4e681a236b570R223) use-cases are listed. Please let me know if I need to clarify anything more there.